### DISCUSSION TITLE
Add (Tuya) AC to Tasmota

### DISCUSSION THREAD

**Discussion by adaasch** [arendst/Tasmota#19236] [Ideas]
Hi,

I reverse engineered the UART protocol of the Tuya WIFI module of my AC and integrated into my home automation via MQTT
Perhaps somebody is interested it: https://github.com/adaasch/AC-hack

It would be nice if Tasmota would directly support the AC. Unfortunately, I do not have good knowledge of the Tasmota code base to do this job efficiently.

**Comment by ferbulous** [comment]
Hi, is there a known list of AC using the tuya module?

**Reply by sfromis** [reply]
I've not seen such, but these two would match....
https://templates.blakadder.com/electriq_IQOOL-SMART12HP
https://templates.blakadder.com/kogan-KAPRA14WFGA

**Reply by adaasch** [reply]
It is a generic Tuya Wifi module -> it does not contain a TuyaMCU.

**Reply by sfromis** [reply]
With a serial protocol, it is the same setup with the main operation being done with another MCU on the main board, even if the protocol may not be the Tuya MCU protocol.

For TuyaMCU-based devices, the generic TYWExx Tuya modules do not "contain" the MCU with the TuyaMCU protocol, as they are effectively a "WiFi modem" add-on to the main board and controller. The "TuyaMCU" is not some specific chip(s), but instead an arbitrary MCU programmed to use the Tuya serial protocol.

**Reply by adaasch** [reply]
I googled the spare part number and came up with these brands:
Kesser
TCL
Pioneer
DAIZUKI

**Reply by ferbulous** [reply]
Thanks! @adaasch  
Not the most common brand sold locally at my place

**Comment by ferbulous** [comment]
> I've not seen such, but these two would match.... https://templates.blakadder.com/electriq_IQOOL-SMART12HP https://templates.blakadder.com/kogan-KAPRA14WFGA

First I've of heard of it, surprisingly come with esp since tuya now use beken chips

**Reply by sfromis** [reply]
Beware that there is very real risk that such devices have silently been upgraded to a newer revision, including a different WiFi module. Of course, in that case swapping the module is not major surgery....

**Comment by Derek-K** [comment]
I just successfully "hacked" my BAC-003 thermostat by replacing the WiFi module with ESP12, flashed with Tasmota and added a bit of scripting (oh, and a minor custom code change) and it works great.

If your thermostat speaks Tuya protocol (it's basically serial protocol, started with 0x55 0xAA) tasmota already have a set of build in command you can use.

Check https://tasmota.github.io/docs/TuyaMCU/ also https://github.com/sillyfrog/Tasmota-Tuya-Helper will help you to discover other commands via ```TuyaSend0```

**Comment by Dobie007** [comment]
I've got the Kesser AC just as adaasch describes it. Apparently it does not have a Tuya MCU in it. 
There is no keep alive on the port.
@adaasch Did you actually flash the TYWE1S module with Tasmota and set that up as serial interface? Till now i am not getting any response from the system?
I flashed the Tasmota light binary, set the config as you show in Github
- Sent SerialConfig 8E1, Baudrate 9600 via console
- Sent Serialsend 5
- waiting for reaction if i switch the AC on - nothing in the console...

Can you give me a hint?

**Reply by sfromis** [reply]
I'd try other bit rates....

**Reply by Dobie007** [reply]
Then i would see something cryptic, but there is nothing. The problem is i dont exactly know which ports of the ESP Chip are connected to the UART1 as they call it in the TYWE1S module.  
And the tasmota documentation says:
Hardware Serial Bridge uses ----  or GPIO13 (Tx) and GPIO15 (Rx) pins of your device.
The ESP pinout has them switched:
![grafik](https://github.com/arendst/Tasmota/assets/59510536/30ec44e2-44a2-4372-beb0-45d23d0ad9b5)
Neither of the setups give me an output..

**Reply by adaasch** [reply]
Yes I put Tasmota on the module, and built a mqtt bridge to control the AC. If you used my settings, maybe Tasmotas own serial output is interfering with the AC chip. Try "SerialLog 0" to deactivate it. To be sure give the AC full shutdown aka remove power.

**Reply by sfromis** [reply]
If you config the expected RX pin as a Counter instead, you'll see the number increment with activity on the pin.

**Reply by barbudor** [reply]
As per http://arduino.esp8266.com/Arduino/versions/2.1.0-rc2/doc/reference.html#serial
It's correctly GPIO15 (TX) and GPIO13 (RX)

So yes, doc is wrong and needs a fix

**Comment by Dobie007** [comment]
Ok, I finally got the first response:  The problem was the way i sent the data:
SerialSend5 BB 00 01 04 02 01 00 BD did not work...  It has to be:
SerialSend5 BB000104020100BD

I flashed the tasmota-lite.bin..
configured GPIO15 (TX) and GPIO13 (RX)
sent: ( via console )
SerialConfig 8E1
Baudrate 9600 
SerialSend5 BB000104020100BD

20:07:20.809 CMD: Grp 0, Cmd 'SERIALSEND', Idx 5, Len 16, Pld -99, Data 'BB000104020100BD'
20:07:20.815 MQT: stat/tasmota_3B510A/RESULT = {"SerialSend":"Done"}
20:07:21.198 MQT: tele/tasmota_3B510A/RESULT = {"SerialReceived":"BB010004370400218A00000000000000006B2B08000000000000000000006BFF4000002A2A2A00008000000000DC0000004440000000001A00000000D2"}

**Comment by Dobie007** [comment]
I tried to use Tasmota scripting now but end up with the next issue: No response from the AC. Using my Picoscope to check the output it reveals that the "8E1" string is not being evaluated correctly in the scriptting.
The script i am using is:

>D 128
res=0
m:trx_a=0 30


>B
res=so(13 15 9600 8E1)
; preset cmd BB 00 01 04 02 01 00 BD
trx_a[1]=0xBB
trx_a[2]=0x00
trx_a[3]=0x01
trx_a[4]=0x04
trx_a[5]=0x02
trx_a[6]=0x01
trx_a[7]=0x00
trx_a[8]=0xBD
 

>S
; send next cmd array with no checksum
print sending command
swa(trx_a 8)

Picoscope output from serial send: (AC answers the command)
![grafik](https://github.com/arendst/Tasmota/assets/59510536/b3b1c02c-4527-4466-903d-ade63c7dcc94)

Picoscope output of the script: (no answer from the AC and Picoscope software says - parity check error)
![grafik](https://github.com/arendst/Tasmota/assets/59510536/9619755d-665d-4e2f-a5fd-517665001822)

I tried to understand the code but didn't understand what is wrong:
![grafik](https://github.com/arendst/Tasmota/assets/59510536/403123cd-d0ba-4e46-a6fd-a8be7d4d5b97)

**Comment by Dobie007** [comment]
Solved see: https://github.com/arendst/Tasmota/issues/19497
Now the work can begin..

**Comment by Dobie007** [comment]
![grafik](https://github.com/arendst/Tasmota/assets/59510536/f80991f1-669d-477b-a2d5-dec75ddff6fe)

Here is the script:

```
>D 128
m:trx_a=0 8
m:rec_a=0 80
m:snd_a=0 40
m:speed=0 6
m:mode=0 5
res=0
rstr=""
p:pwr=0
p:eco=1
p:tbo=0
p:md=5
p:spd=0
Test=""
p:set=20
sw=0
cnt=0
csum=0xFF

>B
so(13 15 9600 8E1)
tper = 5
trx_a={0xBB 0x00 0x01 0x04 0x02 0x01 0x00 0xBD}  
snd_a={0xBB 0x00 0x01 0x03 0x1D 0x00 0x00 0x64 0x01 0x59 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x80 0x18}
speed={0 2 6 3 7 5} 
mode={3 7 2 1 8}

if mqtts>0 {
=#subs
}

#subs
+>Subscribe pwr, cmnd/AC/STATE, Power
+>Subscribe eco, cmnd/AC/STATE, Eco
+>Subscribe tbo, cmnd/AC/STATE, Turbo
+>Subscribe set, cmnd/AC/STATE, Temp
+>Subscribe md, cmnd/AC/STATE, Mode
+>Subscribe spd, cmnd/AC/STATE, Speed

>T
; check serial 
res=sa()
if res>0 {
res=sra(rec_a);
print %res% bytes rec
if (rec_a[1]==0xBB and rec_a[5]==0x37){

pwr = mp(rec_a[8]&0x10 >1 1 <1 0) 
eco = mp(rec_a[8]&0x40 >1 1 <1 0)
tbo = mp(rec_a[8]&0x80 >1 1 <1 0)

md= (rec_a[8]&0x0F)

switch (rec_a[9]&0xF0)
case 0x80 spd= 0
case 0x90 spd= 1 
case 0xC0 spd= 2 
case 0xA0 spd= 3 
case 0xD0 spd= 4 
case 0xB0 spd= 5 
ends
set=(rec_a[9]&0x0F)+16+((rec_a[10]&0x02)*0.25)
}

print send serial
swa(trx_a 8)

>W
bu(pwr "Power on" "Power off") 
bu(eco "ECO on" "ECO off")bu(tbo "Turbo on" "Turbo off") 
sl(16 31 set "16" "Set Temperature  %0set% ℃" "31")
rb(md "Mode" "cooling" "ventilation" "drying" "heating" "auto")
sl(0 5 spd "Auto" "Fan Speed  %0spd%" "Max")

>E  
snd_a[8]=(snd_a[8]&0xFB)|(4*pwr)
snd_a[9]=(snd_a[9]&0xF0)|(mode[md])
snd_a[9]=(snd_a[9]&0xBF)|(0x40*tbo)
snd_a[8]=(snd_a[8]&0x7F)|(0x80*eco)
snd_a[10]=0x6F-set 
snd_a[11]=speed[spd+1]

csum=0
for cnt 1 34 1
csum=csum^snd_a[cnt]
next
snd_a[35]=csum
swa(snd_a 35)
print send new setting
svars-

>J
; send to mqtt
,\n
"Power":%0pwr%,\n
"Eco":%0eco%,\n
"Turbo":%0tbo%,\n
"Temp":%0set%,\n
"Mode":%0md%,\n
"Speed":%0spd%\n_

```
**Compile Tasmota with:**
#ifndef USE_SCRIPT
#define USE_SCRIPT  // adds about 17k flash size, variable ram size
#endif
#ifdef USE_RULES
#undef USE_RULES
#endif 

#define USE_SCRIPT_SERIAL
#define USE_SCRIPT_JSON_EXPORT
#define USE_SCRIPT_WEB_DISPLAY
#define USE_BUTTON_EVENT
#define SUPPORT_MQTT_EVENT

#ifdef USE_DOMOTICZ
#undef USE_DOMOTICZ                              
#endif

**Reply by adaasch** [reply]
Nice Job! So I can get rid of my mqtt bridge :)

**Comment by Dobie007** [comment]
The script still has issues with the different data sources and the timing - so it responds strangely sometimes.
In the meantime i have come over a far better solution: github.com   search for: "WLAN-Klima"
Use "Serial.Swap()" to run it on the original hardware..

**Reply by ilgrank** [reply]
About to buy a Tuya-bases (Argo) Air Conditioner, and was wondering if this code has been addedd to Tasmota or if still a Work In Progress.
Will e.g. this: https://templates.blakadder.com/kogan-KAPRA14WFGA.html be a good starting point?
@Dobie007 : I tried following your instructions but they're a bit beyond my skill level. I've always used Tasmota as it is
Thanks!

**Comment by lNikazzzl** [comment]
I also have an implementation of the same protocol, but based on esphome (works out of the box with home assistant). https://github.com/lNikazzzl/tcl_ac_esphome/tree/master

**Reply by ilgrank** [reply]
@lNikazzzl : checked your project: if I understand correctly I would need to flash ESPHome on the internal MCU?

**Comment by pawol** [comment]
@Dobie007 Would you explain final solution you pointed ? 
I mean WLAN-Klima.
https://github.com/Bro2111/WLAN-Klima

What do you mean by "Use "Serial.Swap()" ?

**Reply by Dobie007** [reply]
I am running Openhab as home automation and for that purpose i don't need a sophisticated Web Interface. I use the Web Interface only to update the software and check what is going on on the local hardware.
Most of the controls are running over MQTT. My local WEB Interface is quite simple:

![grafik](https://github.com/user-attachments/assets/8f6bb10b-ca41-44c3-a10e-1dd1a5ed4c7f)

MQTT:
![grafik](https://github.com/user-attachments/assets/8172801b-87f0-4e7e-930b-ec5fc3866d79)
And here the coding for the TYWE1S hardware interface. (Its nothing near perfect but it does the job for me..)
[Klima04.txt](https://github.com/user-attachments/files/18274325/Klima04.txt)

Regarding the "swap serial": You need to use that function to utilize the serial interface on the secondary pins: 
Serial Comm with GPIO 13,15, 9600 8E1
  }
  //  Verbindung zur Klimaanlage:  
  Serial.println("Switch to SerialKlima");
  Serial.swap();
  delay(200);
  read_AC();
 }

Once you have passed that code the normal serial interface is off and you can use the serial communication with the AC - that is due to the ESP hardware.

**Comment by pawol** [comment]
@Dobie007 
Compiled and I have a few questions...
what does it mean "WT temerature"? cannot translate that abbreviation from german
What is the practical purpose of "FW update" in this case?

Do you have a code supporting Speed, SetTemp functions?

**Reply by Dobie007** [reply]
WT means Wärmetauscher  = "Heat exchanger"

To set the temp. via mqtt post to this topic: "Klima/Funktion" with this payload: {"Set-Temp":21.5 °C}
or "Klima/Funktion" with payload: {"Speed": 3} to set the speed and so on. 
The only thing i didn't care about are the moving fins - It is rather complicated to control and i lost interest.

Firmware update allows to update the software of the ESP  OTA: Over the Air.  It opens a new page asking for the binary file.
